app/src/main/java/org/hitlabnz/sensor_fusion_demo/Compass3D.java

```diff
@@ -0,0 +1,310 @@
+        font = new BitmapFont(); // use libGDX's default Arial font
+        font.getData().setScale(4.0f);
+
+        // for displaying static texts
+        stage = new Stage();
+
+        // Setting up the listener for device volume Up/Down keys press
+        // These keys can be used to switch between the orientation providers
+        Gdx.input.setCatchKey(Input.Keys.VOLUME_UP, true);
+        Gdx.input.setCatchKey(Input.Keys.VOLUME_DOWN, true);
+        Gdx.input.setInputProcessor(new InputAdapter() {
+            @Override
+            public boolean keyUp(int keycode) {
+                switch (keycode) {
+                    case Input.Keys.VOLUME_UP:
+                        p--;
+                        if (p < 0) p = nProviders - 1;
+                        currentOrientationProvider = setCurrentOrientationProvider(p);
+                        break;
+                    case Input.Keys.VOLUME_DOWN:
+                        p++;
+                        if (p == nProviders) p = 0;
+                        currentOrientationProvider = setCurrentOrientationProvider(p);
+                        break;
+                }
+                return true;
+            }
+        });
+
+        // the table will be our parent container for all UI elements
+        table = new Table().top().left();
+        table.setFillParent(true);
+        table.setName("main");
+        //table.setDebug(true);     // useful when we want to visualize the table cells
+
+        stage.addActor(table);
+
+        // create text label to display current FPS
+        label = new Label("FPS:", new LabelStyle(font, Color.YELLOW));
+
+        // add all UI elements to the table, each into a new row
+        table.add(label).top().left().pad(16f);
+        //table.row(); // only if we want to add new UI element to the table in a new row
+
+        // write some logs
+        Gdx.app.log(TAG, "App created!");
+        Gdx.app.log(TAG, "" + Gdx.app.getType());
+        Gdx.app.log(TAG, "GDX version: " + Version.VERSION);
+    }
+
+    @Override
+    public void resize(int width, int height) {
+        // should not happen on Android using the current settings,
+        // manifest file - hardcoded screen portrait mode,
+        // but anyway, on resize we update the camera viewport
+        cam.viewportWidth = width;
+        cam.viewportHeight = height;
+        cam.update();
+
+        stage.getViewport().update(width, height, true);
+    }
+
+    @Override
+    public void render() {
+        // clear screen
+        ScreenUtils.clear(Color.BLACK, true);
+
+        // render the model - our 3D compass
+        modelBatch.begin(cam);
+        modelBatch.render(instance, lights);
+        modelBatch.end();
+
+        // render some text information
+        batch.begin();
+        label.setText("FPS: " + Gdx.graphics.getFramesPerSecond() + "\nGDX version: " + Version.VERSION + "\nCurrent orientation provider:\n" + currentOrientationProvider);
```

Comment:
Is it very complicated to render the text in a nicer resolution with anti-aliasing, that it doesn't look like it just escaped from a 90s computer game?

Reply:
> Is it very complicated to render the text in a nicer resolution with anti-aliasing, that it doesn't look like it just escaped from a 90s computer game?

No, norhing difficult. Similarily as for the buttons, we will need a new truetype font as an asset and load/scale it at the initialization.